paasta_tools/setup_prometheus_adapter_config.py

```diff
@@ -185,9 +236,156 @@ def create_instance_uwsgi_scaling_rule(
+    # templates that are auto-extracted for us. That said: we still need this query to return labels that can be tied
+    # back to k8s objects WITHOUT using label_replace
+    series_query = f"""
+        kube_deployment_labels{{
+            deployment='{deployment_name}',
+            paasta_cluster='{paasta_cluster}',
+            namespace='paasta'
+        }}
+    """
+
+    cpu_usage = f"""
+        sum(
+            irate(
+                container_cpu_usage_seconds_total{{
+                    namespace="paasta",
+                    container="{sanitized_instance_name}",
+                    paasta_cluster="{paasta_cluster}"
+                }}[1m]
+            )
+        ) by (pod, container)
+    """
+
+    cpus_available = f"""
+        sum(
+            container_spec_cpu_quota{{
+                namespace="paasta",
+                container="{sanitized_instance_name}",
+                paasta_cluster="{paasta_cluster}"
+            }}
+            / container_spec_cpu_period{{
+                namespace="paasta",
+                paasta_cluster="{paasta_cluster}"
+            }}
+        ) by (pod, container)
+    """
+
+    # NOTE: we only have Pod names in our container_cpu* metrics, but we can't get a
+    # Deployment from those consistenly due to k8s limitations on certain field lengths
+    # - thus we need to extract this information from the ReplicaSet name (which is made
+    # possible by the fact that our ReplicaSets are named
+    # {{deployment}}-{{10 character hex string}}) so that our query only considers the
+    # service that we want to autoscale - without this we're only filtering by instance
+    # name and these are very much not unique
+    pod_info_join = f"""
+        on (pod) group_left(kube_deployment) label_replace(
+            # k8s:pod:info is an internal recording rule that joins kube_pod_info with
+            # kube_pod_status_phase
+            k8s:pod:info{{
+                created_by_name=~"{deployment_name}.*",
+                created_by_kind="ReplicaSet",
+                namespace="paasta",
+                paasta_cluster="{paasta_cluster}",
+                phase="Running"
+            }},
+            "kube_deployment",
+            "$1",
+            "created_by_name",
+            "(.+)-[a-f0-9]{{10}}"
+        )
+    """
+
+    # get the total usage of all of our Pods divided by the number of CPUs available to
+    # those Pods (i.e., the k8s CPU limit) in order to get the % of CPU used and then add
+    # some labels to this vector
+    average_cpu_usage_per_deployment = f"""
+        avg(
+            (({cpu_usage}) / ({cpus_available})) * {pod_info_join}
+        ) by (kube_deployment)
+    """
+
+    moving_average_cpu_usage_per_deployment = f"""
+        avg_over_time(
+            ({average_cpu_usage_per_deployment})[{moving_average_window}s:]
+        )
+    """
+
```

Comment:
I think instead of `avg_over_time(avg(...) by (kube_deployment))` we want to structure this like the uwsgi query:

```
avg_over_time(
    sum(
        (({cpu_usage} / {cpus_available})) * {pod_info_join}
    )
) / {current_replicas}
```

That is: calculate the desired replicas based on instantaneous load, smooth that over time, and then divide by the non-smoothed current number of replicas.

Otherwise I think we'll see the oscillations that we fixed in https://github.com/Yelp/paasta/pull/2862

Imagine if we've been underscaled by 20% for ~30 minutes at 80 instances (where ideally we should have 100)

1. `avg_over_time()` would give `1.25*setpoint`, and so we'd scale up by 25% to 100 instances.
2. Current load would then be `1*setpoint`. `avg_over_time()` would still show average load of something like `(1 + 29*1.25) / 30 ~= 1.24`, so we'd scale up by 24% to 124 instances.
3. Current load would then be setpoint/1.24. `avg_over_time()` would give `(1/1.24 + 1 + 28*1.25) / 30  ~= 1.22`, so we'd scale up by 22% to 151 instances
4. Current load would then be setpoint/1.24/1.22. `avg_over_time()` would give `(1/1.24/1.22 + 1/1.24 + 1 + 27*1.25) / 30  ~= 1.20`, so we'd scale up by 20% to 181 instances.
5. and so on
6. Eventually avg_over_time() would drop below 1 and we'd start scaling down.
7. By the time we scale back down to 100, `avg_over_time()` is substantially below 1 (because we've been overscaled for a long time) and so we'll overshoot in the other direction